### PR TITLE
[receiver/awscloudwatchmetricsreceiver] Add new receiver

### LINKED ISSUE #15667: New component: AWS CloudWatch metrics receiver
### The purpose and use-cases of the new component

Collect metrics from AWS Cloudwatch metrics and parse them in to OTel pmetric. Some metrics on AWS, such as Lambda memory usage etc, are  only available from CloudWatch — users may want them to be put in one platform for analysis such as Prometheus with Grafana.

### Example configuration for the component

```
receivers:
    awscloudwatchmetrics:
        region: <aws_region>
        profile: <aws_profile> # optional if omitted will use default for current if the AWS credential environment variables are not set
        polling_interval: 60
        metrics:
          named: 
             - namespace: AWS/ALB
                metric_name: RequestCount
             - namespace: AWS/ALB
                metric_name: 500Count
```

### Telemetry data types supported

Metrics

### Is this a vendor-specific component?

- [X] This is a vendor-specific component
- [ ] If this is a vendor-specific component, I am proposing to contribute this as a representative of the vendor.

### Sponsor (optional)

_No response_

### Additional context

Functionality is similar to AWSContainerMetrics receiver, maybe extend that instead and make it more generic?

### PR DESCRIPTION
**Pull request by lewis262626**
**Description:** New AWS CloudWatch metrics receiver initial PR

**Link to tracking Issue:** #15667 

**Testing:** Unit tests for component config and factory

PR supersedes: #19218

### CODE REVIEW

**Comment by lewis262626** [comment]
cc @jpkrohling @kovrus

File: .github/ISSUE_TEMPLATE/feature_request.yaml
Comment (atoulme) [inline .github/ISSUE_TEMPLATE/feature_request.yaml:1]:
```suggestion
name: Feature request
```

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (kovrus) [inline receiver/awscloudwatchmetricsreceiver/config.go:69]: Maybe we shall add the information about the limit to the error message. You've done that for `errTooManyDimensions`, so let's add it here too.

**Review by kovrus — APPROVED** [review]
LGTM

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go]: Fixed in [8ab9607](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/19429/commits/8ab9607cd8a44fe14c0e9f55646914d57d3ee8fb)

**Comment by lewis262626** [comment]
Unsure why CI is failing. Logs are saying it's failing on a Jaeger test 🤔

**Comment by atoulme** [comment]
It’s failing on checks too, please take a look

**Comment by lewis262626** [comment]
@atoulme I fixed the issue with `make goporto`, but the `jaeger` issue seems to be nothing related to my commits. The logs say something is already listening on that port:

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4387063396/jobs/7685532139#step:9:520

**Comment by atoulme** [comment]
It's a flaky test: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9113

**Comment by lewis262626** [comment]
> It's a flaky test: #9113

Ok, hopefully a re-run fixes it

**Comment by atoulme** [comment]
Please add the new module to versions.yaml with `Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml`

**Comment by lewis262626** [comment]
> Please add the new module to versions.yaml with `Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml`

Cheers, fixed.

**Review by atoulme — APPROVED** [review]

**Comment by lewis262626** [comment]
Can someone with write permissions merge this PR please?

**Review by matej-g — COMMENTED** [review]
Thanks for this @lewis262626!

I left a few questions / suggestions. It seems like the receiver does not implement any CloudWatch API polling yet? Is this meant to provide just bare bones for the receiver? 

(Sorry if that's just my unfamiliarity with process, not sure what is the required minimum viable functionality for a receiver to be merged. I was expecting an initial version to already be capable of receiving metrics).

File: receiver/awscloudwatchmetricsreceiver/receiver.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/receiver.go:39]: Will fix, cheers

File: receiver/awscloudwatchmetricsreceiver/README.md
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/README.md:13]: Typo, should be v2. Thanks!

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go:122]: Makes sense, will add an extra error and update tests

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go:28]: It depends, you can get get high precision AWS metrics that have a 1s period. I think for AWS metrics it's 5m, so let's do that. Cheers

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go:69]: I think this can be removed as I have implemented the split logic already

**Comment by matej-g** [comment]
@lewis262626 thanks for the explanation, let's get this in! ❤️

**Review by codeboten — COMMENTED** [review]
Please resolve/address the remaining comments and we can get this merged!

**Review by jpkrohling — COMMENTED** [review]
dependabot conflict solved, but apparently, there are a few unhandled comments. I'll go over this once the comments are taken care of.

**Comment by lewis262626** [comment]
> @lewis262626 thanks for the explanation, let's get this in! ❤️ 

I have some free time later this week. Will resolve the outstanding comments and rebase

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go]: Fixed in 5bcaf6eed472b209affdfad37bfd1914508c5126

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go]: Fixed in df9aefd825d1267621fff213c0ef700c556ffc78

File: receiver/awscloudwatchmetricsreceiver/README.md
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/README.md]: Fixed in df9aefd825d1267621fff213c0ef700c556ffc78

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go]: Fixed in df9aefd825d1267621fff213c0ef700c556ffc78

File: receiver/awscloudwatchmetricsreceiver/receiver.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/receiver.go]: Fixed in 92c82032c7

**Comment by TylerHelmuth** [comment]
@jpkrohling please give final review as the sponsor.

File: .github/CODEOWNERS
Comment (atoulme) [inline .github/CODEOWNERS:162]:
```suggestion
receiver/awscloudwatchmetricsreceiver/                   @open-telemetry/collector-contrib-approvers @jpkrohling @kovrus
```

Given that you're not a member of OpenTelemetry yet, we cannot list you as codeowner. Please see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20868 as background.

An easy way to go about this is to remove your name for the time being, merge this in and use me and the sponsor of the component @jpkrohling as sponsors to get you added as member of the OpenTelemetry organization.

**Review by atoulme — COMMENTED** [review]
Please make sure to do the following:
* [x] rebase with latest main
* [x] run `make gendependabot`
* [x] update all version numbers to 0.77.0

That should be enough to get you ready for merge.

**Comment by lewis262626** [comment]
@atoulme That should be all done. Appreciate your help :)

Great to get this approved and merged in this week

**Review by matej-g — APPROVED** [review]
Let's get this in 💪 , thanks @lewis262626!

File: receiver/awscloudwatchmetricsreceiver/README.md
Comment (codeboten) [inline receiver/awscloudwatchmetricsreceiver/README.md:7]: Doesn't have to be in this PR, but please follow up with an update to auto-generate the status information

**Review by codeboten — APPROVED** [review]

File: receiver/awscloudwatchmetricsreceiver/config_test.go
Comment (jpkrohling) [inline receiver/awscloudwatchmetricsreceiver/config_test.go:43]: Is there a real value that would have meaning to users? Tests are documentation as well, using real-world examples to help people understand the code under test.

File: receiver/awscloudwatchmetricsreceiver/config.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config.go:100]: I can see both angles tbh, probably worth a follow-up issue to see if some advice can go in to the `CONTRIBUTING.md` or the opentelemetry docs

File: receiver/awscloudwatchmetricsreceiver/config_test.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config_test.go:217]:
```suggestion
							Name:  ":P99",
```

File: receiver/awscloudwatchmetricsreceiver/README.md
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/README.md:112]: I'm just following what the AWS docs say. I can make it explicit that it's the number of metrics you define in the receiver configuration, and not how many API calls the receiver makes.

File: receiver/awscloudwatchmetricsreceiver/config_test.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config_test.go:43]: Agreed. I'll fix this

File: receiver/awscloudwatchmetricsreceiver/config_test.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config_test.go:154]: Great spot. Fixed

File: receiver/awscloudwatchmetricsreceiver/README.md
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/README.md:112]: I think it's worth just deleting that sentence. The pricing changes fairly frequently and varies between regions. It's the user's decision to weigh up pricing etc. None of the other receivers such as Splunk etc have pricing. 

Large customers could also have discounted pricing etc.

I can follow up with an issue to see if a custom collector metric around number of api calls and number of requested metrics would be useful?

File: receiver/awscloudwatchmetricsreceiver/config_test.go
Comment (lewis262626) [reply receiver/awscloudwatchmetricsreceiver/config_test.go:43]: I'll fix the remaining issues and rebase tomorrow

**Comment by lewis262626** [comment]
Should all be fixed. Appreciate a final review @jpkrohling

**Comment by dmitryax** [comment]
@lewis262626 please take a look at the failing linter

**Comment by lewis262626** [comment]
> @lewis262626 please take a look at the failing linter

Fixed the linter issue

File: receiver/awscloudwatchmetricsreceiver/config_test.go
Comment (jpkrohling) [reply receiver/awscloudwatchmetricsreceiver/config_test.go:143]: Is this fixed? I still see something odd here. I won't block the PR because of this, but please double-check on a follow-up PR.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**
